syntax = "proto3nbt";
package Game.NbtChunk;

message ArmorItem {
}

message Attribute {
  double base = 1 [label = "Base"];
  string name = 2 [label = "Name"];
}

message Memories {
}

message Brain {
  Memories memories = 1;
}

message HandItem {
}

message Entity {
  float absorption_amount = 1 [label = "AbsorptionAmount"];
  int16 air = 2 [label = "Air"];
  repeated float armor_drop_chances = 3 [label = "ArmorDropChances"];
  repeated ArmorItem armor_items = 4 [label = "ArmorItems"];
  repeated Attribute attributes = 5 [label = "Attributes"];
  Brain brain = 6 [label = "Brain"];
  int8 can_pick_up_loot = 7 [label = "CanPickUpLoot"];
  int16 death_time = 8 [label = "DeathTime"];
  int8 explosion_radius = 9 [label = "ExplosionRadius"];
  float fall_distance = 10 [label = "FallDistance"];
  int8 fall_flying = 11 [label = "FallFlying"];
  int16 fire = 12 [label = "Fire"];
  int16 fuse = 13 [label = "Fuse"];
  repeated float hand_drop_chances = 14 [label = "HandDropChances"];
  repeated HandItem hand_items = 15 [label = "HandItems"];
  float health = 16 [label = "Health"];
  int32 hurt_by_timestamp = 17 [label = "HurtByTimestamp"];
  int16 hurt_time = 18 [label = "HurtTime"];
  int8 invulnerable = 19 [label = "Invulnerable"];
  int8 left_handed = 20 [label = "LeftHanded"];
  repeated double motion = 21 [label = "Motion"];
  int8 on_ground = 22 [label = "OnGround"];
  int8 persistence_required = 23 [label = "PersistenceRequired"];
  int32 portal_cooldown = 24 [label = "PortalCooldown"];
  repeated double pos = 25 [label = "Pos"];
  repeated float rotation = 26 [label = "Rotation"];
  ints uuid = 27 [label = "UUID"];
  string id = 28;
  int8 ignited = 29;
}

message Heightmaps {
  longs motion_blocking = 1 [label = "MOTION_BLOCKING"];
  longs motion_blocking_no_leaves = 2 [label = "MOTION_BLOCKING_NO_LEAVES"];
  longs ocean_floor = 3 [label = "OCEAN_FLOOR"];
  longs world_surface = 4 [label = "WORLD_SURFACE"];
}

message PaletteItem {
  string name = 1 [label = "Name"];
  map properties = 2 [label = "Properties"];
}

message Section {
  int8 y = 1 [label = "Y"];
  bytes block_light = 2 [label = "BlockLight"];
  bytes sky_light = 3 [label = "SkyLight"];
  longs block_states = 4 [label = "BlockStates"];
  repeated PaletteItem palette = 5 [label = "Palette"];
}

message References {
  longs bastion_remnant = 1;
  longs buried_treasure = 2;
  longs desert_pyramid = 3;
  longs fortress = 4;
  longs jungle_pyramid = 5;
  longs mineshaft = 6;
  longs monument = 7;
  longs nether_fossil = 8;
  longs ocean_ruin = 9;
  longs pillager_outpost = 10;
  longs shipwreck = 11;
  longs swamp_hut = 12;
}

message StartId {
  string id = 1;
}

message Starts {
  StartId mineshaft = 1;
  StartId ruined_portal = 2;
  StartId stronghold = 3;
}

message Structures {
  References references = 1 [label = "References"];
  Starts starts = 2 [label = "Starts"];
}

message Level {
  ints biomes = 1 [label = "Biomes"];
  repeated Entity entities = 2 [label = "Entities"];
  Heightmaps heightmaps = 3 [label = "Heightmaps"];
  int64 inhabited_time = 4 [label = "InhabitedTime"];
  int64 last_update = 5 [label = "LastUpdate"];
  /* repeated repeated <unknown> post_processing = 6 [label = "PostProcessing"]; */
  repeated Section sections = 7 [label = "Sections"];
  string status = 8 [label = "Status"];
  Structures structures = 9 [label = "Structures"];
  /* repeated <unknown> tile_entities = 10 [label = "TileEntities"]; */
  /* repeated <unknown> tile_ticks = 11 [label = "TileTicks"]; */
  int8 is_light_on = 12 [label = "isLightOn"];
  int32 x_pos = 13 [label = "xPos"];
  int32 z_pos = 14 [label = "zPos"];
}

message Chunk {
  Level level = 1 [label = "Level"];
}
